File: java/beer-song/src/main/java/BeerSong.java
import java.util.stream.Collectors;
import java.util.stream.IntStream;

class BeerSong {
    String sing(int startBeers, int numVerses) {
        return IntStream.iterate(startBeers, beers -> beers > startBeers - numVerses, beers -> beers - 1)
                        .mapToObj(this::sing)
                        .collect(Collectors.joining());
    }

    String singSong() {
        return sing(99, 100);
    }

    private String sing(int beers) {
        if (beers > 0) {
            return String.format("%1$d bottle%2$s of beer on the wall, %1$d bottle%2$s of beer.\n" +
                    "Take %3$s down and pass it around, %4$s bottle%5$s of beer on the wall.\n\n",
                    beers,
                    beers > 1 ? "s" : "",
                    beers != 1 ? "one" : "it",
                    beers != 1 ? String.valueOf(beers - 1) : "no more",
                    beers != 2 ? "s" : "");
        } else {
            return "No more bottles of beer on the wall, no more bottles of beer.\n" +
                    "Go to the store and buy some more, 99 bottles of beer on the wall.\n\n";
        }
    }
}

File: r/beer-song/beer-song.R
lyrics <- function(first, last) {
  paste0(sapply(c(first:last), verse), collapse = "\n")
}

verse <- function(number) {
  num.str <- paste0(number)
  paste0(switch(num.str, "0" = "No more bottles", "1" = "1 bottle", paste0(number, " bottles")),
         " of beer on the wall, ",
         switch(num.str, "0" = "no more bottles", "1" = "1 bottle", paste0(number, " bottles")),
         " of beer.\n",
         switch(num.str, "0" = "Go to the store and buy some more",
                         "1" = "Take it down and pass it around",
                         "Take one down and pass it around"),
         ", ",
         switch(num.str, "0" = "99 bottles", "1" = "no more bottles", "2" = "1 bottle", paste0(number - 1, " bottles")),
         " of beer on the wall.\n")
}
